File: web/types/agent.ts
// Backend agent types - must match src/models/enums.py
export type AgentType = "SimpleAgent" | "LLMAgent" | "RAGAgent" | "ToolAgent" | "FullAgent" | "RouterAgent" | "OrchestratorAgent";

// RouterAgent configuration - maps intent categories to target agents
export interface RouterConfig {
  routing_table: Record<string, RoutingEntry>;
  confidence_threshold?: number;
}

export interface RoutingEntry {
  target_agent: string;
  description?: string;
}

export interface AgentRole {
  title: string;
  expertise: string[];
  personality: string[];  // API expects array
  communication_style: string;
}

export interface AgentGoal {
  objective: string;
  success_criteria: string[];
  constraints: string[];
}

export interface AgentInstructions {
  steps: string[];
  rules: string[];
  prohibited: string[];  // Must match backend field name
  output_format: string;
}

export interface AgentExample {
  input: string;
  output: string;
}

export interface LLMConfig {
  provider: "openai" | "anthropic";
  model: string;
  temperature: number;
}

export interface ToolReference {
  tool_id: string;
}

export interface SkillConfig {
  skill_id: string;
  enabled: boolean;
  parameters: Record<string, unknown>;
  priority: number;
}

export interface KnowledgeBaseConfig {
  collection_name: string;
  embedding_model?: string;
  top_k?: number;
  similarity_threshold?: number;
}

// API expects lowercase
export type SafetyLevel = "low" | "standard" | "high" | "maximum";

export interface SafetyConfig {
  level: SafetyLevel;
  blocked_topics: string[];
  blocked_patterns: string[];
}

export interface AgentCreate {
  id: string;
  name: string;
  description: string;
  agent_type: AgentType;
  role: AgentRole;
  goal: AgentGoal;
  instructions: AgentInstructions;
  examples: AgentExample[];
  llm_config: LLMConfig;
  knowledge_base?: KnowledgeBaseConfig;
  tools: ToolReference[];
  skills?: SkillConfig[];
  safety: SafetyConfig;
  // RouterAgent specific
  router_config?: RouterConfig;
  // OrchestratorAgent specific
  orchestrator_config?: OrchestratorConfig;
}

export interface Agent {
  id: string;
  name: string;
  description: string;
  agent_type: AgentType;
  created: boolean;
}

// Full agent details for editing
export interface AgentDetail {
  id: string;
  name: string;
  description: string;
  agent_type: AgentType;
  role: AgentRole;
  goal: AgentGoal;
  instructions: {
    steps: string[];
    rules: string[];
    prohibited: string[];
    output_format: string | null;
  };
  examples: AgentExample[];
  llm_config: LLMConfig | null;
  knowledge_base: KnowledgeBaseConfig | null;
  tools: ToolReference[];
  skills?: SkillConfig[];
  routing_table: Record<string, string> | null;
  orchestrator_config: OrchestratorConfig | null;
  safety: SafetyConfig;
}

export interface OrchestratorConfig {
  mode: string;
  available_agents: AgentReference[];
  workflow_definition: string | null;
  default_aggregation: string;
  max_parallel: number;
  max_depth: number;
  allow_self_reference: boolean;
}

export interface AgentReference {
  agent_id: string;
  alias: string | null;
  description: string | null;
}

export interface AgentListResponse {
  agents: Agent[];
  total: number;
}

export interface WorkflowRequest {
  agent_id: string;
  user_input: string;
  conversation_history?: { role: "user" | "assistant"; content: string }[];
  session_id?: string;
}

// Tool call result returned in metadata
export interface ToolCallResult {
  tool: string;
  args: Record<string, unknown>;
  result: {
    success: boolean;
    output: unknown;
    error?: string;
  };
}

// Matches backend ExecuteAgentResponse schema
export interface WorkflowResponse {
  content: string;
  agent_id: string;
  agent_type: AgentType;
  success: boolean;
  error: string | null;
  metadata: {
    model?: string;
    usage?: {
      prompt_tokens: number;
      completion_tokens: number;
      total_tokens: number;
    };
    tool_calls?: ToolCallResult[];
    iterations?: number;
    validation_retries?: number;
    finish_reason?: string;
    loop_detected?: boolean;
    execution_time_ms?: number;
    // Orchestrator specific
    agent_results?: Record<string, unknown>[];
    mode?: string;
  };
  workflow_id: string | null;

  // Clarification fields - for interactive follow-up questions
  requires_clarification?: boolean;
  clarification_question?: string;
  clarification_options?: string[];
}

export type WorkflowStatus = "RUNNING" | "COMPLETED" | "FAILED" | "CANCELLED";

export interface WorkflowStatusResponse {
  workflow_id: string;
  status: WorkflowStatus;
  result?: {
    content: string;
    success: boolean;
  };
}

// AI Generation types
export interface GenerateAgentRequest {
  name: string;
  context?: string;
}

export interface GenerateAgentResponse {
  description: string;
  role: AgentRole;
  goal: AgentGoal;
  instructions: {
    steps: string[];
    rules: string[];
    prohibited: string[];
    output_format: string | null;
  };
  examples: AgentExample[];
  suggested_agent_type: AgentType;
}

// ============================================================================
// Agent Overview Tab Types
// ============================================================================

export type TimeRange = "24h" | "7d" | "30d" | "90d";

export interface AgentStats {
  agent_id: string;
  time_range: TimeRange;

  // Performance metrics
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  success_rate: number; // 0.0 to 1.0
  avg_latency_ms: number;
  p95_latency_ms: number;

  // Cost metrics
  total_cost_cents: number;
  total_tokens: number;

  // Trends (percentage change vs previous period)
  executions_trend: number;
  success_trend: number;
  latency_trend: number;
  cost_trend: number;
}

export interface ExecutionSummary {
  id: string;
  status: "completed" | "failed";
  timestamp: string; // ISO format
  duration_ms: number;
  input_preview: string | null;
  output_preview: string | null;
  error_type: string | null;
  error_message: string | null;
  total_tokens: number;
  total_cost_cents: number;
}

export interface AgentExecutionsResponse {
  executions: ExecutionSummary[];
  total: number;
  has_more: boolean;
}

export interface UsageDataPoint {
  timestamp: string; // ISO format
  executions: number;
  successful: number;
  failed: number;
  avg_latency_ms: number;
  total_cost_cents: number;
}

export interface AgentUsageHistory {
  data: UsageDataPoint[];
  granularity: "hour" | "day";
  time_range: TimeRange;
}

// Full execution details with metadata
export interface ExecutionDetail {
  id: string;
  agent_id: string;
  agent_type: AgentType;
  timestamp: string;
  status: "completed" | "failed";
  duration_ms: number;
  input_preview: string | null;
  output_preview: string | null;
  total_tokens: number;
  total_cost_cents: number;
  llm_calls_count: number;
  tool_calls_count: number;
  error_type: string | null;
  error_message: string | null;
  workflow_id: string | null;
  execution_metadata: ExecutionMetadata | null;
}

export interface ExecutionMetadata {
  model?: string;
  usage?: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
  tool_calls?: ToolCallDetail[];
  agent_results?: AgentCallResult[];
  iterations?: number;
  mode?: string;
  // Add other potential fields
  [key: string]: unknown;
}

export interface ToolCallDetail {
  tool: string;
  args: Record<string, unknown>;
  result: {
    success: boolean;
    output: unknown;
    error?: string;
  };
}

export interface AgentCallResult {
  agent_id: string;
  content: string;
  success: boolean;
  error?: string;
  metadata?: Record<string, unknown>;
}

File: src/api/schemas/agent.py
"""API schemas for agents."""
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict, Field

from src.models.enums import AgentType, SafetyLevel


class AgentRoleSchema(BaseModel):
    """API schema for agent role."""

    title: str
    expertise: List[str] = []
    personality: List[str] = []
    communication_style: str = "professional"


class AgentGoalSchema(BaseModel):
    """API schema for agent goal."""

    objective: str
    success_criteria: List[str] = []
    constraints: List[str] = []


class AgentInstructionsSchema(BaseModel):
    """API schema for agent instructions."""

    steps: List[str] = []
    rules: List[str] = []
    prohibited: List[str] = []
    output_format: Optional[str] = None


class AgentExampleSchema(BaseModel):
    """API schema for agent example."""

    input: str
    output: str


class LLMConfigSchema(BaseModel):
    """API schema for LLM config."""

    provider: str = "anthropic"
    model: str = "claude-3-sonnet-20240229"
    temperature: float = 0.7
    max_tokens: int = 1024


class KnowledgeBaseConfigSchema(BaseModel):
    """API schema for knowledge base config."""

    collection_name: str
    embedding_model: str = "text-embedding-3-small"
    top_k: int = 5
    similarity_threshold: Optional[float] = None


class ToolConfigSchema(BaseModel):
    """API schema for tool config."""

    tool_id: str
    enabled: bool = True
    requires_confirmation: bool = False


class SkillConfigSchema(BaseModel):
    """API schema for skill config."""

    skill_id: str
    enabled: bool = True
    parameters: Dict[str, Any] = {}
    priority: int = 0


class SafetyConfigSchema(BaseModel):
    """API schema for safety config."""

    level: SafetyLevel = SafetyLevel.STANDARD
    blocked_topics: List[str] = []
    blocked_patterns: List[str] = []


class AgentReferenceSchema(BaseModel):
    """API schema for agent reference in orchestrator."""

    agent_id: str
    alias: Optional[str] = None
    description: Optional[str] = None


class OrchestratorConfigSchema(BaseModel):
    """API schema for orchestrator config."""

    mode: str = "llm_driven"  # llm_driven, workflow, hybrid
    auto_discover: bool = False  # If True, automatically discover all available agents
    exclude_agent_types: List[str] = ["OrchestratorAgent"]  # Agent types to exclude when auto_discover is True
    exclude_agent_ids: List[str] = []  # Specific agent IDs to exclude when auto_discover is True
    available_agents: List[AgentReferenceSchema] = []  # Ignored if auto_discover is True
    workflow_definition: Optional[str] = None
    default_aggregation: str = "all"  # first, all, vote, merge, best
    max_parallel: int = 5
    max_depth: int = 3
    allow_self_reference: bool = False


class CreateAgentRequest(BaseModel):
    """Request to create an agent."""

    id: str = Field(..., min_length=1, max_length=100)
    name: str = Field(..., min_length=1, max_length=200)
    description: str = ""
    agent_type: AgentType
    role: AgentRoleSchema
    goal: AgentGoalSchema
    instructions: AgentInstructionsSchema = Field(default_factory=AgentInstructionsSchema)
    examples: List[AgentExampleSchema] = []
    llm_config: Optional[LLMConfigSchema] = None
    knowledge_base: Optional[KnowledgeBaseConfigSchema] = None
    tools: List[ToolConfigSchema] = []
    skills: List[SkillConfigSchema] = []
    routing_table: Optional[Dict[str, str]] = None
    orchestrator_config: Optional[OrchestratorConfigSchema] = None
    safety: SafetyConfigSchema = Field(default_factory=SafetyConfigSchema)


class AgentResponse(BaseModel):
    """Response containing agent summary."""

    id: str
    name: str
    description: str
    agent_type: AgentType
    created: bool = True
    warnings: List[str] = Field(
        default_factory=list,
        description="Warnings about potential issues (e.g., hallucination risks)"
    )

    class Config:
        from_attributes = True


class AgentDetailResponse(BaseModel):
    """Response containing full agent details for editing."""

    id: str
    name: str
    description: str
    agent_type: AgentType
    role: AgentRoleSchema
    goal: AgentGoalSchema
    instructions: AgentInstructionsSchema
    examples: List[AgentExampleSchema] = []
    llm_config: Optional[LLMConfigSchema] = None
    knowledge_base: Optional[KnowledgeBaseConfigSchema] = None
    tools: List[ToolConfigSchema] = []
    skills: List[SkillConfigSchema] = []
    routing_table: Optional[Dict[str, str]] = None
    orchestrator_config: Optional[OrchestratorConfigSchema] = None
    safety: SafetyConfigSchema

    class Config:
        from_attributes = True


class AgentListResponse(BaseModel):
    """Response containing list of agents."""

    agents: List[AgentResponse]
    total: int


class GenerateAgentRequest(BaseModel):
    """Request to generate agent configuration with AI."""

    name: str = Field(..., min_length=1, max_length=200)
    context: Optional[str] = Field(None, max_length=500, description="Optional context about what the agent should do")


class GenerateAgentResponse(BaseModel):
    """Response containing AI-generated agent configuration."""

    description: str
    role: AgentRoleSchema
    goal: AgentGoalSchema
    instructions: AgentInstructionsSchema
    examples: List[AgentExampleSchema] = []
    suggested_agent_type: AgentType


# ============================================================================
# Agent Overview Tab - Stats & Executions
# ============================================================================


class AgentStatsResponse(BaseModel):
    """Response containing agent performance statistics."""

    agent_id: str
    time_range: str  # "24h", "7d", "30d", "90d"

    # Performance metrics
    total_executions: int
    successful_executions: int
    failed_executions: int
    success_rate: float  # 0.0 to 1.0
    avg_latency_ms: float
    p95_latency_ms: float

    # Cost metrics
    total_cost_cents: int
    total_tokens: int

    # Trends (percentage change vs previous period)
    executions_trend: float  # +12.5 means 12.5% increase
    success_trend: float
    latency_trend: float
    cost_trend: float


class ExecutionSummary(BaseModel):
    """Summary of a single execution."""

    id: str
    status: str  # "completed", "failed"
    timestamp: str  # ISO format
    duration_ms: int
    input_preview: Optional[str] = None
    output_preview: Optional[str] = None
    error_type: Optional[str] = None
    error_message: Optional[str] = None
    total_tokens: int = 0
    total_cost_cents: int = 0


class AgentExecutionsResponse(BaseModel):
    """Response containing list of agent executions."""

    executions: List[ExecutionSummary]
    total: int
    has_more: bool


class ExecutionDetailResponse(BaseModel):
    """Full execution details including metadata."""

    id: str
    agent_id: str
    agent_type: str
    timestamp: str  # ISO format
    status: str  # "completed", "failed"
    duration_ms: int
    input_preview: Optional[str] = None
    output_preview: Optional[str] = None
    total_tokens: int = 0
    total_cost_cents: int = 0
    llm_calls_count: int = 0
    tool_calls_count: int = 0
    error_type: Optional[str] = None
    error_message: Optional[str] = None
    workflow_id: Optional[str] = None
    execution_metadata: Optional[Dict[str, Any]] = None

    model_config = ConfigDict(extra="ignore")


class UsageDataPoint(BaseModel):
    """Single data point for usage chart."""

    timestamp: str  # ISO format
    executions: int
    successful: int
    failed: int
    avg_latency_ms: float
    total_cost_cents: int


class AgentUsageHistoryResponse(BaseModel):
    """Response containing usage history for charts."""

    data: List[UsageDataPoint]
    granularity: str  # "hour" or "day"
    time_range: str
